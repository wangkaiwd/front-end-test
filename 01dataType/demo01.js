// js数据类型
// 基本数据类型（值类型）： 1. string 2. number 3. boolean 4. null 5. undefined 7. Symbol(es6新增) 8. Bigint(新增)
// 复杂数据类型（引用类型）： 6. object: Array Function Regexp Date Math

let a = { x: 100 };
let fn = () => {
  console.log('object key is a function');
};
let array = [1, 2, 3];
let obj = {
  0: 100,
  true: 'wk',
};
obj[a] = 'another obj';
obj[fn] = 'function';
obj[array] = 'array';
// {
//   '0': 100,
//   true: 'wk',
//   '[object Object]': 'another obj',
//   "() => {\n  console.log('object key is a function');\n}": 'function',
//   '1,2,3': 'array'
// }
console.log(obj);

// 对象的key会转换为字符串(调用toString方法)
// ({0:100,true: 'wk'}).toString() => [object Object]
// [1,2,3].toString() => 1,2,3
// (() => {console.log('object key is a function')}).toString() => '() => {\n  console.log('object key is a function');\n}'

// 一道面试题
// const a = {
//   x: 100,
// };
// const b = {
//   y: 200,
// };
// const obj = {};
// obj[a] = 'aaa';
// obj[b] = 'bbb';
//
// console.log('obj', obj);
