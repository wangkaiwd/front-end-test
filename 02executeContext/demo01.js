// 编译器(把代码解析成为浏览器可以看的懂的结构)：
//    词法解析
//    AST抽象语法树
//    构建出浏览器能够执行的代码
// 引擎(V8 webkit内核)
//    变量声明提升
//    作用域 闭包
//    变量对象
//    堆栈内存
//    GO/VO/AO/EC/ECStack
//    ...

// let a = 12;
// let b = a;
// b = 13;
// console.log('a', a);
//
// let a = {
//   n: 12,
// };
//
// let b = a;
// b.n = 13;
// console.log(a.n);
