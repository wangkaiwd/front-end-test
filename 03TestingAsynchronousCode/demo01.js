export const testCallback = (cb) => {
  setTimeout(() => {
    cb('peanut butter');
  }, 2000);
};
export const testPromise = () => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve('peanut butter');
    }, 3000);
  });
};

export const testPromiseError = () => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      reject('error');
    }, 3000);
  });
};
