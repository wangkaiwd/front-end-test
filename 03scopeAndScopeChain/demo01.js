function A (y) {
  let x = 2;
  function B (z) {
    console.log(x + y + z);
  }
  return B;
}
let C = A(2);
C(3);
