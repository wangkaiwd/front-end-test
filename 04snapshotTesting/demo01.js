export const generateConfig = () => {
  return {
    server: 'http://localhost',
    port: 8080,
  };
};

export const anotherGenerateConfig = () => {
  return {
    server: 'http://localhost',
    port: 3002,
  };
};
