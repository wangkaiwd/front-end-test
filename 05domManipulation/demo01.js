const addDivToBody = () => {
  const div = document.createElement('div');
  document.body.appendChild(div);
};

export {
  addDivToBody
};
