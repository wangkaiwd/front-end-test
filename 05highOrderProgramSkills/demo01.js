// 在没有模块化的情况下，进行多人协作开发(单例模式)

// 假设有俩个业务模块：商品模块，物流模块

// 为了防止变量干扰，每个都使用自执行函数保护内部的变量

// 没有框架的时候可以这样用：
// 单例设计模式(最原始的模块化思想)：
//  1. 利用闭包机制将元素包起来
//  2. 将外界想要访问的内容通过return返回
// 商品模块
const shopModule = (() => {
  const getList = () => { //获取商品列表

  };

  const getCategory = () => { // 获取商品分类

  };
  return {
    // 这里通过对象返回，如果还有内容需要供外界使用，可以继续为对象提供属性
    getCategory,
  };
})();

// 新闻模块
const newsModule = (() => {
  const getList = () => { // 获取物流信息列表

  };
  // 在这里也想使用商品模块的getCategory:
  //    1. 将getCategory放到全局作用域中，则所有的模块都可以直接使用。但是这样会加大全局变量被污染的概率
  //    2. 在商品模块将要用到的内容返回，并用一个全局变量接收
  shopModule.getCategory();
  return {};
})();
