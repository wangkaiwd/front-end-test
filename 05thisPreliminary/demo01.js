var x = 10; // window.x => 10 , 注意：let , const 不会将变量挂载到window中
const obj = {
  x: 20,
  fn () {
    console.log(this.x);
  },
};

const fn = obj.fn;

// 假设页面中有id为box的dom元素
const $box = document.getElementById('box');
$box.x = 30;

fn(); // 10

obj.fn(); // 20

$box.addEventListener('click', function () {
  obj.fn(); // 20
});

$box.addEventListener('click', obj.fn); // 30
