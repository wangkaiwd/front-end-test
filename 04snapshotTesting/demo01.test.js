import { anotherGenerateConfig, generateConfig } from './demo01';

test('test snapshot', () => {
  // 第一次运行会在测试文件旁边生成一个快照文件，
  // 之后再次运行测试文件，会生成新的快照文件与第一次生成的进行对比，如果不一致就会报错
  // 如果之后内容为正确的内容，那么修改快照，否则需要还原更改
  // watch状态下通过w: show more，然后选择命令进行更新快照
  expect(generateConfig()).toMatchSnapshot();
});

// 同一个文件中的多个快照出现更改时，可以通过i进入交互模式，对每一个快照进行分别更改
// 如果想要直接更新所有快照，可以通过u来进行直接更新
test('test another snapshot', () => {
  expect(anotherGenerateConfig()).toMatchSnapshot();
});
