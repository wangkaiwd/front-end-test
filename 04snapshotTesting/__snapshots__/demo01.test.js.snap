// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test another snapshot 1`] = `
Object {
  "port": 3002,
  "server": "http://localhost",
}
`;

exports[`test snapshot 1`] = `
Object {
  "port": 8080,
  "server": "http://localhost",
}
`;
